# Exploring eBay Car Sales Data

## Importing Libraries

In [1]:
import numpy as np 
import pandas as pd 

## Loading our dataset

In [2]:
autos = pd.read_csv("autos.csv",encoding = "Latin-1")

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

## Knowing more about our dataset 

In [4]:
print(f"info is: '\n' {autos.info()}")
print("\n")
autos.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

**The dataset contains 20 columns, most of which are strings,from the above display we can see that we have around 50000 row in which some of the columns are missing values and also we have multiple datatypes**

In [5]:
prev_col_index = autos.columns
prev_col_index.unique()

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

**Previous column names seems like they are using the camel syntax but we prefer the python snake syntax**

In [6]:
def cleaner(val):
    val = val.replace("yearOfRegistration","registration_year")
    val = val.replace("monthOfRegistration","registration_month")
    val = val.replace("notRepairedDamage","unrepaired_damage")
    val = val.replace("dateCreated","ad_created")
    val = val.replace("dateCrawled","date_crawled")
    val = val.replace("lastSeen","last_seen")
    return val
new_col_lead = []
for value in prev_col_index:
    clean_val = cleaner(value)
    new_col_lead.append(clean_val)


**A bit of cleaning**

In [7]:
autos.columns=new_col_lead

In [8]:
new_col_lead

['date_crawled',
 'name',
 'seller',
 'offerType',
 'price',
 'abtest',
 'vehicleType',
 'registration_year',
 'gearbox',
 'powerPS',
 'model',
 'odometer',
 'registration_month',
 'fuelType',
 'brand',
 'unrepaired_damage',
 'ad_created',
 'nrOfPictures',
 'postalCode',
 'last_seen']

**Now this is better**

## we turned the column indexes from Camel Case to snake case which is much easier to use in python because of syntax 

## price and odometer are numeric values stord as strings 

In [9]:
autos.describe(include="all")

date_crawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  registration_year  gearbox       powerPS  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuelType       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrOfPictures    postalCode            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

**We can clearly see that from the dataframe description**

**Our initial observations**:

There are a number of text columns where all (or nearly all) of the values are the same:
<ul>
    <li> seller
    <li> offer_type
</ul>

In [10]:
autos["seller"].value_counts()

seller
privat        49999
gewerblich        1
Name: count, dtype: int64

In [11]:
autos["offerType"].value_counts()

offerType
Angebot    49999
Gesuch         1
Name: count, dtype: int64

In [12]:
autos.dtypes 

date_crawled          object
name                  object
seller                object
offerType             object
price                 object
abtest                object
vehicleType           object
registration_year      int64
gearbox               object
powerPS                int64
model                 object
odometer              object
registration_month     int64
fuelType              object
brand                 object
unrepaired_damage     object
ad_created            object
nrOfPictures           int64
postalCode             int64
last_seen             object
dtype: object

In [13]:
autos["price"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

**Let's do a bit of cleaning to restore our numerical column**

In [14]:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].astype(int)

In [15]:
autos["seller"].head()

0    privat
1    privat
2    privat
3    privat
4    privat
Name: seller, dtype: object

In [16]:
autos["offerType"].head()

0    Angebot
1    Angebot
2    Angebot
3    Angebot
4    Angebot
Name: offerType, dtype: object

In [17]:
autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

**This column should also be cleaned to restore it back to it's numerical status**

In [18]:
autos["odometer"] = autos["odometer"].str.replace(",","")
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].astype(int)

**Renaming our columns so that we could undertand the unit's of measurment in the future**

In [19]:
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)

**Let's check our largest odometer reading we have**

In [20]:
max_odom=autos["odometer_km"].max()
max_odom

150000

**Minimum reading**

In [21]:
min_odom=autos["odometer_km"].min()
min_odom

5000

In [22]:
print(autos["odometer_km"].describe())
print(f"maximum odometer reading is {max_odom} and minimum odometer reading is {min_odom}")

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
maximum odometer reading is 150000 and minimum odometer reading is 5000


**Let's check out our price column**

In [23]:
print(autos["price"].describe())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


**Doesn't look quite right look at the huge differnce between our minimum and maximum values let's do something about it**

In [24]:
low, high = autos["price"].quantile([0.1,0.9])
mask = autos["price"].between(low,high)
autos = autos[mask]
max_price=autos["price"].max()
min_price=autos["price"].min()
print(f"maximum odometer reading is {max_price} and minimum odometer reading is {min_price}")

maximum odometer reading is 14000 and minimum odometer reading is 500


**That's better we used the 10th percentile as our lower limit and the 90th percentile as our upper limit to filter out our outliers**

## date_crawled

In [25]:
autos["date_crawled"].head()

0    2016-03-26 17:47:46
1    2016-04-04 13:38:56
2    2016-03-26 18:57:24
3    2016-03-12 16:58:10
4    2016-04-01 14:38:50
Name: date_crawled, dtype: object

In [26]:
date_crawled=autos["date_crawled"].str[:10]
date_crawled.value_counts(normalize=True, dropna=False).sort_index()

date_crawled
2016-03-05    0.025721
2016-03-06    0.013993
2016-03-07    0.036353
2016-03-08    0.033863
2016-03-09    0.032867
2016-03-10    0.033290
2016-03-11    0.033091
2016-03-12    0.038096
2016-03-13    0.015388
2016-03-14    0.036901
2016-03-15    0.034137
2016-03-16    0.029754
2016-03-17    0.031622
2016-03-18    0.012674
2016-03-19    0.034037
2016-03-20    0.037946
2016-03-21    0.037548
2016-03-22    0.032742
2016-03-23    0.032568
2016-03-24    0.028933
2016-03-25    0.031622
2016-03-26    0.032917
2016-03-27    0.030726
2016-03-28    0.034884
2016-03-29    0.033116
2016-03-30    0.033440
2016-03-31    0.031373
2016-04-01    0.032892
2016-04-02    0.035083
2016-04-03    0.038295
2016-04-04    0.036577
2016-04-05    0.013072
2016-04-06    0.003262
2016-04-07    0.001220
Name: proportion, dtype: float64

**Earliest date crawled was `2016-03-05` and the latest date was `2016-04-07`, on `2016-04-03` lots of dates were crawled**

## ad_created

In [27]:
autos["ad_created"].head()

0    2016-03-26 00:00:00
1    2016-04-04 00:00:00
2    2016-03-26 00:00:00
3    2016-03-12 00:00:00
4    2016-04-01 00:00:00
Name: ad_created, dtype: object

In [28]:
date_created=autos["ad_created"].str[:10]
date_created.value_counts(normalize=True, dropna=False).sort_index()

ad_created
2015-08-10    0.000025
2015-09-09    0.000025
2015-11-10    0.000025
2015-12-05    0.000025
2015-12-30    0.000025
                ...   
2016-04-03    0.038594
2016-04-04    0.036975
2016-04-05    0.011777
2016-04-06    0.003312
2016-04-07    0.001096
Name: proportion, Length: 71, dtype: float64

**The earliest created date was `2015-08-10` and the latest date was `2016-04-07`**

## last_seen

In [29]:
autos["last_seen"].head()

0    2016-04-06 06:45:54
1    2016-04-06 14:45:08
2    2016-04-06 20:15:37
3    2016-03-15 03:16:28
4    2016-04-01 14:38:50
Name: last_seen, dtype: object

In [30]:
last_seen=autos["last_seen"].str[:10]
last_seen.value_counts(normalize=True, dropna=False).sort_index()

last_seen
2016-03-05    0.001195
2016-03-06    0.004557
2016-03-07    0.005627
2016-03-08    0.007694
2016-03-09    0.010159
2016-03-10    0.010881
2016-03-11    0.012898
2016-03-12    0.025621
2016-03-13    0.009586
2016-03-14    0.012674
2016-03-15    0.016284
2016-03-16    0.017031
2016-03-17    0.028983
2016-03-18    0.007694
2016-03-19    0.016209
2016-03-20    0.021239
2016-03-21    0.021612
2016-03-22    0.021911
2016-03-23    0.019048
2016-03-24    0.020368
2016-03-25    0.019421
2016-03-26    0.017156
2016-03-27    0.016309
2016-03-28    0.021737
2016-03-29    0.022160
2016-03-30    0.024849
2016-03-31    0.024102
2016-04-01    0.023629
2016-04-02    0.025397
2016-04-03    0.025472
2016-04-04    0.025123
2016-04-05    0.121483
2016-04-06    0.215104
2016-04-07    0.126787
Name: proportion, dtype: float64

**The earliest date was `2016-03-05` , the latest `2016-04-07`** 

## Let's check out registration_year column

In [31]:
autos["registration_year"].describe()

count    40162.000000
mean      2004.141477
std         75.928363
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2007.000000
max       9999.000000
Name: registration_year, dtype: float64

## count    40162.000000
## mean      2004.141477
 ## min       1000.000000 **_seems to be an outlier_**
## max       9999.000000 **_seems to be an outlier_**

**The min regestration year and max regestration year are clearly in paradox hence we need to specify our range**

In [32]:
lowest,highest =1900,2016
mask_year_reg = autos["registration_year"].between(lowest,highest)
autos = autos[mask_year_reg]
autos["registration_year"].describe()

count    38455.000000
mean      2002.550280
std          6.489912
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2007.000000
max       2016.000000
Name: registration_year, dtype: float64

## Regestration year is much better and the years are much more realistic 

In [33]:
autos["brand"].head()

0       peugeot
1           bmw
2    volkswagen
3         smart
4          ford
Name: brand, dtype: object

In [34]:
brand_sorted=autos["brand"].value_counts(normalize=True).sort_values(ascending=False)[:20]

In [35]:
brand_index=brand_sorted.index
brand_index

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'smart', 'mazda', 'nissan',
       'citroen', 'toyota', 'hyundai', 'volvo', 'honda', 'mitsubishi'],
      dtype='object', name='brand')

## I have chosen to work with the top 20 car brands with the highest percentage occurnce 

In [36]:
brand_repo = {}
for brand in brand_index:
    price= autos.loc[autos["brand"]==brand,"price"]
    price_mean = price.mean()
    brand_repo[brand]=price_mean
milege_repo = {}   
for brand in brand_index:
    distance_coverd = autos.loc[autos["brand"]==brand,"price"]
    distance_mean = distance_coverd.mean()
    milege_repo[brand]=distance_mean
    

In [37]:
print(brand_repo)

{'volkswagen': 4161.386629660915, 'opel': 3023.230696798493, 'bmw': 5368.535356006851, 'mercedes_benz': 4903.687042253521, 'audi': 5333.960674157303, 'ford': 3194.3713318284426, 'renault': 2505.3902569593147, 'peugeot': 3110.638110488391, 'fiat': 3100.6160714285716, 'seat': 3864.9586776859505, 'skoda': 5389.680882352941, 'smart': 3590.8012232415904, 'mazda': 3661.2109500805154, 'nissan': 3918.8406040268455, 'citroen': 3554.6195286195284, 'toyota': 4443.058287795992, 'hyundai': 4779.701711491442, 'volvo': 3774.5625, 'honda': 3802.876923076923, 'mitsubishi': 2993.5862068965516}


## There different price range for differnt cars in our top 20 cars from 5368 to 2505

In [38]:
brand_repo_series = pd.Series(brand_repo)
brand_repo_series
milege_repo_series= pd.Series(milege_repo)
milege_repo_series


volkswagen       4161.386630
opel             3023.230697
bmw              5368.535356
mercedes_benz    4903.687042
audi             5333.960674
ford             3194.371332
renault          2505.390257
peugeot          3110.638110
fiat             3100.616071
seat             3864.958678
skoda            5389.680882
smart            3590.801223
mazda            3661.210950
nissan           3918.840604
citroen          3554.619529
toyota           4443.058288
hyundai          4779.701711
volvo            3774.562500
honda            3802.876923
mitsubishi       2993.586207
dtype: float64

In [39]:
brand_repo_dataframe = pd.DataFrame(brand_repo_series,milege_repo_series,columns=["Average_price"])

In [40]:
brand_repo_dataframe

Average_price
4161.386630            NaN
3023.230697            NaN
5368.535356            NaN
4903.687042            NaN
5333.960674            NaN
3194.371332            NaN
2505.390257            NaN
3110.638110            NaN
3100.616071            NaN
3864.958678            NaN
5389.680882            NaN
3590.801223            NaN
3661.210950            NaN
3918.840604            NaN
3554.619529            NaN
4443.058288            NaN
4779.701711            NaN
3774.562500            NaN
3802.876923            NaN
2993.586207            NaN